### ISSUE TITLE
Tags/Labels friendly stats system for Envoy

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#20289]
*Title*: *Tags-friendly stats system for Envoy*

*Description*:

Here is a further discussion of #20220. Envoy provides powerful stats system which is well designed and carefully optimised. This system provides first-class support for statsd. Regular expression-based extraction of tags and output in prometheus format were added later.

The current method of using regular expressions to extract tags is very flexible and powerful. But if there are some irregular identifiers (such as including variable length dots), the regular expression will be very complicated.  And the regular expression for extracting tags can only be configured through bootstrap.

So may be we can try to do some enhancements to current stats system to get better support to the tags/labels stats.

This PR is used to record the disccusion.

[optional *Relevant Links*:]

Related comments: https://github.com/envoyproxy/envoy/issues/20220#issuecomment-1061986614

Initial idea about the enhancement: https://docs.google.com/document/d/1Y5bHKcwiFE_cueH3obYfQ59RCwz9688ZC9mNGOYW_hM/edit?usp=sharing

**Comment by wbpcode** [comment]
cc @mattklein123 @ggreenway @jmarantz

**Comment by mattklein123** [comment]
Sorry for the delay. I took a look at the doc and in general this makes sense to me, though we have to sort out the details. Whatever we do I think we have to continue to support both flat names and taggified names for a while. 

IMO the best way to do this is in line with what you have proposed. Effectively for new call sites allow both the flat and taggified variants to be specified, so we can avoid having to add new regex. Is that basically what you have in mind?

**Comment by wbpcode** [comment]
> Sorry for the delay. I took a look at the doc and in general this makes sense to me, though we have to sort out the details. Whatever we do I think we have to continue to support both flat names and taggified names for a while. 
> 
> IMO the best way to do this is in line with what you have proposed. Effectively for new call sites allow both the flat and taggified variants to be specified, so we can avoid having to add new regex. Is that basically what you have in mind?

Yeah, it is what I want. A stats system that can support flat flattened name and also flexible tags.
I am still thinking about the implementation details. May be the first step, we should design new scope api to support additional tag annotations.

**Comment by ggreenway** [comment]
I was thinking about this again, after doing some more work on fixing broken tag-extraction rules. @wbpcode do you still hope to implement this at some point?

**Comment by wbpcode** [comment]
yeah. I still wanna it. But I didn't find a way that perfectly compatible with the current behavior.
